gpflow/base.py

```diff
@@ -307,10 +305,36 @@ def _run_op(a, *args, **kwargs):
 Parameter._OverloadAllOperators()
 tf.register_tensor_conversion_function(Parameter, lambda x, *args, **kwds: x.read_value())
 
+TensorLike = Union[_TensorLikeExternalTypes, Parameter]
 
-def _cast_to_dtype(value: VariableData, dtype: Optional[DType] = None) -> tf.Tensor:
+
+class Module(tf.Module):
+    @property
+    def parameters(self) -> Tuple[Parameter, ...]:
+        return tuple(self._flatten(predicate=_IS_PARAMETER))
+
+    @property
+    def trainable_parameters(self) -> Tuple[Parameter, ...]:
+        return tuple(self._flatten(predicate=_IS_TRAINABLE_PARAMETER))
+
+    def _repr_html_(self) -> str:
+        from .utilities import tabulate_module_summary
+
+        return tabulate_module_summary(self, tablefmt="html")
+
+    def _repr_pretty_(self, p: pretty.RepresentationPrinter, cycle: bool) -> None:
```

Comment:
I'm generally happy to add types but in this instance I don't think it's worth adding the IPython dependency and import. Could mypy handle it as
```suggestion
    def _repr_pretty_(self, p: "IPython.lib.pretty.RepresentationPrinter", cycle: bool) -> None:
```
?
If not, I would forgo the type
```suggestion
    def _repr_pretty_(self, p, cycle: bool) -> None:
```
you can still add it in a docstring if you want.

Reply:
(here's how forward references work)[https://www.python.org/dev/peps/pep-0484/#forward-references]. in particular, putting types as strings only works if the expression makes sense once the file is loaded - i.e. we'd still need to import that class. The answer, if we're going to omit it, is to use `Any` and comment why. We can discuss that in the thread above

Reply:
It's not Any, though; I agree that adding types can help, but adding type hints that aren't actually correct I think is worse than not having any type whatsoever.

Reply:
Though in any case, it's a private method that will only internally be called by IPython, so we might as well point to the IPython docs and omit type hints here altogether imo.

Reply:
you're right that typing it explicitly as `Any` isn't great cos it gives the impression that's what it is, when it's not. What I'm trying to avoid is leaving types out cos then it's not clear if that's been done out of omission or some other reason. See my other comment for my latest thought

Reply:
solved: see above